pdata/ptrace/traces.go

```diff
@@ -46,5 +46,10 @@ func (ms Traces) SpanCount() int {
 
 // ResourceSpans returns the ResourceSpansSlice associated with this Metrics.
 func (ms Traces) ResourceSpans() ResourceSpansSlice {
-	return newResourceSpansSlice(&ms.getOrig().ResourceSpans)
+	return newResourceSpansSlice(&ms.getOrig().ResourceSpans, internal.GetTracesState(internal.Traces(ms)))
+}
+
+// Shared returns a Traces that shares underlying resources with ms.
+func (ms Traces) Shared() Traces {
```

Comment:
Few things to consider here:
1. Why return a new Traces vs changing the current Traces? Since you don't do a deep copy even the current Traces is shared.
2. I did propose Shared as a name but I am thinking if ReadOnly is easier for people to understand.
3. Do we need a IsShared()? If we have that we can get rid of the "Capabilities" in the consumers, and the consumers that need to mutate will check if IsShared first thing then will do a Clone if they are "mutable" and object is shared/readonly. Not ideal solution but is better than Capabilities I believe.
4. We should look a bit at the interaction of this concept with the Request.Traces(). That Request.Traces() may have to change to better fit this. Just something to make sure we check before final design.
5. I also want to think if we should also implement a ref-counting (separate of sharable) and optimize our allocs by using pools of allocated pointers. The reason I am mentioning this is just I want you to think if that will still be possible and the two concepts can live separately and be clearly defined independently.

Reply:
> 3. Do we need a IsShared()? If we have that we can get rid of the "Capabilities" in the consumers, and the consumers that need to mutate will check if IsShared first thing then will do a Clone if they are "mutable" and object is shared/readonly. Not ideal solution but is better than Capabilities I believe.

I don't think we need it. As I mentioned in the issue description, in order to get rid of the capabilities, we can just introduce another method `func (t Traces) Exclusive() Traces` which will return either the same object. if it's already exclusive or a copy if it's shared. That's why I thought it's better to have `func (t Traces) Shared() Traces` for consistency, replying to your (1) comment. If you agree with this, we need another name for that method if we switch to `ReadOnly` (2).

> 4. We should look a bit at the interaction of this concept with the Request.Traces(). That Request.Traces() may have to change to better fit this. Just something to make sure we check before final design.

I think I addressed that in this PR. The state can be set both on `Traces` and `ExportRequest`, so if you convert from one another, it's preserved. Unmarshalling gives a request with the Exclusive state.

> 5. I also want to think if we should also implement a ref-counting (separate of sharable) and optimize our allocs by using pools of allocated pointers. The reason I am mentioning this is just I want you to think if that will still be possible and the two concepts can live separately and be clearly defined independently.

I don't think it'll conflict with this implementation, but can dig deeper. I think this is something that we can do along with the copy-on-write after the 1.0 release. Based on our latest discussion, the pointer-based CoW solution can be a significant change for the clients, so, if it works out, I think it should go to 2.0. That way, we can provide a shim to support both during the transition period. WDYT?

Reply:
Maybe we still need `IsShared` in case components want to do different things based on that.

Reply:
I kind of don't understand why you need the return value even in the Exclusive case, why not cloning internally the referenced data? The reason I am asking because if I have an "exclusive" pdata and I mark it as "shared" with the current API, I can end up with one object exclusive and one shared that point to the same data:

```golang
td := ptrace.NewTraces()
std := td.Shared()

td.ResourceSpans().AppendEmpty() // works fine
std.ResourceSpans().AppendEmpty() // panics

// Hence if I pass `td` to a pipeline, and `std` to a different one I get UB.
```

Reply:
This makes sense. I liked `Shared()` because we don't have to introduce a method with a pointer receiver. Linter might be complaining that we have both methods on both value and pointer receivers, but I  think that's okay. I don't believe we can migrate other methods to pointers without affecting end users.

Reply:
After giving it a second thought, I don't think we are safe from this UB even with this solution. I still can do this:

```golang
td := ptrace.NewTraces()
tds := td
td.MarkShared()

// td is shared, but tds is still exclusive
```

I implemented your suggestion before realizing it. Now, I think `Shared()` is still a cleaner option.

Reply:
As discussed in another thread, I migrated the state to a pointer, so the concern above is not valid anymore.

Reply:
@bogdandrutu speaking about `ReadOnly` vs `Shared`. What do you think of the following API:

```
func (ms [Logs|Traces|Metrics]) MarkAsReadOnly()
func (ms [Logs|Traces|Metrics]) IsReadOnly()
```

This can be enough for now, users will be able to use `CopyTo` to get writable. Or, optionally, we could add the following later:

```
func (ms [Logs|Traces|Metrics]) MarkAsMutable()
```

Reply:
Not a fan of the "MarkAs". What about just `ReadOnly` or `MakeReadOnly` since will make more sense `MakeMutable`.

Reply:
Not a fan of the "MarkAs". What about just `ReadOnly` or `MakeReadOnly` since will make more sense `MakeMutable`.

Reply:
`MakeReadOnly` sounds good to me. I believe it has to involve an action as it's a mutating function. 

Then we should have `IsMutable` and `MakeMutable`? I think `IsMutable` is something that users want to know essentially, so it's better than writing `!logs.IsReadOnly()`.

Reply:
`ToReadOnly` or `MakeReadOnly` both are good for the moment. Let's do the `Is` and `Mutable` in the next PR.

Reply:
Another option, could be to have `MarkReadOnly` and `Clone`? Then `Clone` can return a value.

Reply:
`MarkReadOnly` is also a good option. Not sure I understand why do we need `Clone`? We used to have it and replaced it with `CopyTo`

Reply:
Let's go with `MarkReadOnly` for the moment, and will discuss in the next PR the next change.

Reply:
Let's go with `MarkReadOnly` for the moment, and will discuss in the next PR the next change.

Reply:
Applied `MarkReadOnly` 👍